### ISSUE TITLE
Pod templates are lost after building BuildConfig

### ISSUE THREAD

**Issue by major** [openshift/jenkins-sync-plugin#236]
In my OpenShift environment, I have a few BuildConfigs that prepare container images that Jenkins uses as slave nodes. I need some volumes mounted when Jenkins builds its nodes and I set those up via configmaps using the guide shown here:

https://docs.openshift.com/container-platform/3.6/using_images/other_images/jenkins.html#using-the-jenkins-kubernetes-plug-in-to-run-jobs

However, if I rebuild one of the containers, Jenkins will not use that imagestream again for that container. I can remove my configmap and re-apply it to fix the issue:

```
$ oc delete -f nodes-configmap.yml
$ oc apply -f nodes-configmap.yml
```

After that step, Jenkins will create Kubernetes pods using that imagestream again. Is it possible to have Jenkins be able to use that imagestream when the image is rebuilt?

**Comment by gabemontero** [comment]
Hey @major - going to need some clarifications, but it *sounds* like you are trying to manage a single pod template from both a config map and image stream/ image stream tag ... if so, that is not supported

You can either
a) create a pod template from a config map, where you have to specify a specific image, but you can set any of the pod template config fields, including volume mounts

b) create a pod template from an image stream / image stream tag, where we'll grab the appropriate version of an image from those entities, but we only give you a standard pod template def with a limited set of config options, and you cannot do things like mount additional volumes

You cannot manage a single pod template from both a config map and image stream

If that does now sound like what you are doing, the please provide yaml for
1) the build configs in question
2) the image streams who receive the output from the build config
3) the config maps that set up the volumes, etc.

**Comment by gabemontero** [comment]
OK that yaml helped clarify what you are trying to do.

*Seems* like it should work, but with what you are doing, from the sync plugin perspective, to only thing to verify is that 

`<image>docker-registry.default.svc:5000/mynamespace/centos7-builder</image>`

Is correctly translated to the pod template you see in the configuration for the k8s plugin.

Assuming it is, then we have moved beyond this plugin, to either the k8s plugin doing something unexpected, or the platform doing something unexpected.

Here are the steps you need to verify:
1) after the `oc start-build centos7-builder` completes, verify the image stream centos7-builder getting update, and in particular its latest tag, with a new version of `docker-registry.default.svc:5000/mynamespace/centos7-builder:latest` ... and if you docker pull that image from the internal registry, is it updated?
2) does `docker-registry.default.svc:5000` to the same IP in all the different node(s) in play here?
3) Assuming everything is running in `mynamespace`, run `oc get events -w -n mynamespace` when you run a jenkins pipeline that uses `centos7-builder` ... do you see the kubelet pulling the image down?
4) Run `oc get pods <pod created from centos7-builder> -o yaml ... is the AlwaysPull flag set?  Does the image field look right?

Based on the answers to those questions, we should hopefully know to either tweak your set up, or with other repo to report your situation to.

**Comment by gabemontero** [comment]
On Tue, May 1, 2018 at 3:20 PM, Ben Parees <notifications@github.com> wrote:

> @major <https://github.com/major> what do you mean by "However, if I
> rebuild one of the containers, Jenkins will not use that imagestream again
> for that container. " ?
>
> What happens when you launch a new jenkins job that uses the slave label,
> after pushing a new version of the image? I'm not clear if the issue is
> that no slave pod is being created, or if the slave pod that gets created
> is not using the new image. (it sounds like the former, but i want to be
> sure)
>
I totally took it as the later :-) ... so sure, clarifying that aspect
sounds prudent

**Comment by major** [comment]
@bparees Sorry, I worded that pretty horribly. ;) Here are the steps that cause a problem for me:

1. Load the ImageStream/BuildConfig YAML. OpenShift automatically starts a build of the image.
2. Once the build completes, I load the ConfigMap.
3. I tell Jenkins to run a job on a node with one of the labels from my ConfigMap.
4. The job runs normally and everything is fine.
5. I update the repo in GitHub that OpenShift uses to build my container.
6. I go to OpenShift and start a new build for that container.
7. When the build is complete, I start a job in Jenkins on a node with a label that corresponds to the image I just built.
8. Jenkins sits around forever and says I have no PodTemplate for that label.
9. I remove the ConfigMap (`oc delete -f ...`) and add it again (`oc apply -f ...`).
10. Jenkins waits a few minutes and then asks OpenShift to build a slave node using the label I requested.

One thing I noticed is that when I remove/apply the ConfigMap, I see lines in the Jenkins log showing that the Pod Template was removed and added again. However, when I rebuild the BuildConfig, I don't see those lines in the Jenkins log.

**Comment by gabemontero** [comment]
OK, that was my understanding, though you dropped in a couple of more clues @major that could narrow the focus

On you last point, you don't need to see the remove/add message each time, assuming the pod template definition is still present after your first run.

I'll adjust my list from https://github.com/openshift/jenkins-sync-plugin/issues/236#issuecomment-385742303 based on the additional information you provided:

1) provide me the entire jenkins log
2) still have a run of `oc get events -w` for the namespace(s) in question ... where ever the builds for the image are occurring, wherever jenkins is running ... provide that output
3) confirm *BEFORE* you do the `oc delete -f ...` and `oc apply -f ...`, but after you rebuild the slave image, that the pod template config is correct ... do an `oc rsync <pod name>:/var/lib/jenkins <local dir` and then grab the config.xml from that local dir and provide that as well.

It sounds more like a k8s plugin issue (unless something unexpected is happening the the pod template definition).

What version of the openshift jenkins image and k8s plugin are you running with?

ok, that's it ... thanks

**Comment by gabemontero** [comment]
reopen if you are still messing with this @major and can provide the data requested at https://github.com/openshift/jenkins-sync-plugin/issues/236#issuecomment-385985595

thanks

**Comment by etoews** [comment]
For anyone stumbling across this in the future, this is being resolved over in #307.
